rasa/core/policies/ted_policy.py

```diff
@@ -0,0 +1,642 @@
+from rasa.utils.tensorflow.transformer import TransformerEncoder
+from rasa.utils.tensorflow.models import RasaModel
+from rasa.utils.tensorflow.model_data import RasaModelData, FeatureSignature
+from rasa.utils.tensorflow.constants import (
+    LABEL,
+    HIDDEN_LAYERS_SIZES,
+    TRANSFORMER_SIZE,
+    NUM_TRANSFORMER_LAYERS,
+    NUM_HEADS,
+    MAX_SEQUENCE_LENGTH,
+    BATCH_SIZES,
+    BATCH_STRATEGY,
+    EPOCHS,
+    RANDOM_SEED,
+    RANKING_LENGTH,
+    LOSS_TYPE,
+    SIMILARITY_TYPE,
+    NUM_NEG,
+    EVAL_NUM_EXAMPLES,
+    EVAL_NUM_EPOCHS,
+    NEGATIVE_MARGIN_SCALE,
+    REGULARIZATION_CONSTANT,
+    SCALE_LOSS,
+    USE_MAX_NEG_SIM,
+    MAX_NEG_SIM,
+    MAX_POS_SIM,
+    EMBEDDING_DIMENSION,
+    DROP_RATE_DIALOGUE,
+    DROP_RATE_LABEL,
+    DROP_RATE_ATTENTION,
+    KEY_RELATIVE_ATTENTION,
+    VALUE_RELATIVE_ATTENTION,
+    MAX_RELATIVE_POSITION,
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class TEDPolicy(Policy):
+    """Transformer Embedding Dialogue (TED) Policy.
+
+    The policy used in our paper https://arxiv.org/abs/1910.00486.
+    """
+
+    SUPPORTS_ONLINE_TRAINING = True
+
+    # default properties (DOC MARKER - don't remove)
+    defaults = {
+        # nn architecture
+        # a list of hidden layers sizes before dialogue and action embedding layers
+        # number of hidden layers is equal to the length of this list
+        HIDDEN_LAYERS_SIZES: {DIALOGUE: [], LABEL: []},
+        # number of units in transformer
+        TRANSFORMER_SIZE: 128,
+        # number of transformer layers
+        NUM_TRANSFORMER_LAYERS: 1,
+        # max sequence length
+        MAX_SEQUENCE_LENGTH: 256,
+        # number of attention heads in transformer
+        NUM_HEADS: 4,
+        # if true use key relative embeddings in attention
+        KEY_RELATIVE_ATTENTION: False,
+        # if true use key relative embeddings in attention
+        VALUE_RELATIVE_ATTENTION: False,
+        # max position for relative embeddings
+        MAX_RELATIVE_POSITION: None,
+        # training parameters
+        # initial and final batch sizes:
+        # batch size will be linearly increased for each epoch
+        BATCH_SIZES: [8, 32],
+        # how to create batches
+        BATCH_STRATEGY: "balanced",  # 'sequence' or 'balanced'
+        # number of epochs
+        EPOCHS: 1,
+        # set random seed to any int to get reproducible results
+        RANDOM_SEED: None,
+        # embedding parameters
+        # dimension size of embedding vectors
+        EMBEDDING_DIMENSION: 20,
+        # the type of the similarity
+        NUM_NEG: 20,
+        # flag if minimize only maximum similarity over incorrect labels
+        SIMILARITY_TYPE: "auto",  # 'auto' or 'cosine' or 'inner'
+        # the type of the loss function
+        LOSS_TYPE: "softmax",  # 'softmax' or 'margin'
+        # number of top actions to normalize scores for softmax loss_type
+        # set to 0 to turn off normalization
+        RANKING_LENGTH: 10,
```

Comment:
```suggestion
        ACTION_RANKING_LENGTH: 10,
```

Reply:
We reuse this variable in other models, would not rename it.

Reply:
> We reuse this variable in other models, would not rename it.

Does this mean that this constant has different meanings for the other models?